mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSessionHolder.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.orm.session.impl;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.transaction.Synchronization;
+
+import org.hibernate.Transaction;
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.search.mapper.orm.common.EntityReference;
+import org.hibernate.search.mapper.pojo.work.spi.PojoIndexingPlan;
+
+public class HibernateOrmSearchSessionHolder implements Serializable {
+
+	private static final String SESSION_PROPERTY_KEY = "hibernate.search.session";
+
+	private static final Map<Transaction, HibernateOrmSearchSessionHolder> holderPerClosedSessionTransaction =
+			new ConcurrentHashMap<>();
+
+	// Public for tests only
+	public static int staticMapSize() {
+		return holderPerClosedSessionTransaction.size();
+	}
+
+	public static HibernateOrmSearchSessionHolder get(SessionImplementor session) {
+		return get( session, true );
+	}
+
+	public static HibernateOrmSearchSessionHolder get(SessionImplementor session, boolean createIfMissing) {
+		HibernateOrmSearchSessionHolder holder = (HibernateOrmSearchSessionHolder)
+				session.getProperties().get( SESSION_PROPERTY_KEY );
+		if ( holder != null ) {
+			return holder;
+		}
+		boolean closedSessionAndTransaction = session.isClosed() && session.isTransactionInProgress();
+		if ( closedSessionAndTransaction ) {
+			// This can happen when using JTA.
+			// Only do this as a fallback: if somehow the session holder was added to the session properties
+			// before the session was closed, we definitely want to use it and avoid the static map.
+			holder = holderPerClosedSessionTransaction.get( session.accessTransaction() );
+		}
```

Comment:
This is not critical either, the code looks good in any case, but maybe we can introduce 2 sub classes: one for JTA and the other for the non-JTA case.

Reply:
This is a static method, no polymorphism here.

Also, we will not always end up in this branch, even with JTA: if a `SearchSession` was retrieved before the commit, we'll proceed with the "normal" behavior (which is better).